extensions/amp-viewer-integration/0.1/findtext.js

```diff
@@ -0,0 +1,451 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {computedStyle} from '../../../src/style';
+
+/**
+ * Simple implementation of CircularBuffer.
+ * Exported for test only.
+ */
+export class CircularBuffer {
+  constructor(max) {
+    this.max_ = max;
+    this.buff_ = [];
+    this.next_ = 0;
+  }
+
+  add(item) {
+    this.buff_[this.next_] = item;
+    this.next_ = (this.next_ + 1) % this.max_;
+  }
+
+  /**
+   * @param {number} index The index of an element to get.
+   */
+  get(index) {
+    if (this.buff_.length >= this.max_) {
+      index = (this.next_ + index) % this.max_;
+    }
+    return this.buff_[index];
+  }
+}
+
+/**
+ * TextPos is a pointer to a character in a Text node.
+ * Exported for test only.
+ */
+export class TextPos {
```

Comment:
Is this abstraction necessary? You can just use a basic @typedef.

Reply:
Done